src/main/java/com/teragrep/cfe_35/router/ParallelTargetRouting.java

```diff
@@ -49,95 +49,145 @@
+    private final RoutingConfig routingConfig;
+    private final RelpClientFactory clientFactory;
+
+    private final RelpFrameFactory relpFrameFactory;
+    private final FragmentFactory fragmentFactory;
+    private final FragmentStub fragmentStub;
+
+    public ParallelTargetRouting(
+            RoutingConfig routingConfig,
+            MetricRegistry metricRegistry,
+            RelpClientFactory clientFactory
+    ) throws IOException {
+        this.routingConfig = routingConfig;
+        this.clientFactory = clientFactory;
+        this.relpFrameFactory = new RelpFrameFactory();
+        this.fragmentFactory = new FragmentFactory();
+        this.fragmentStub = new FragmentStub();
 
-    public ParallelTargetRouting(RoutingConfig routingConfig, MetricRegistry metricRegistry) throws IOException {
         this.totalRecords = metricRegistry.counter(name(ParallelTargetRouting.class, "totalRecords"));
         this.totalBytes = metricRegistry.counter(name(ParallelTargetRouting.class, "totalBytes"));
 
+    }
+
+    private Map<String, CompletableFuture<RelpClient>> createClients() {
+        LOGGER.debug("in createClients");
+        Map<String, CompletableFuture<RelpClient>> outputs = new HashMap<>();
         Map<String, TargetConfig> configMap = routingConfig.getTargetConfigMap();
+
         for (Map.Entry<String, TargetConfig> entry : configMap.entrySet()) {
             String targetName = entry.getKey();
             TargetConfig targetConfig = entry.getValue();
             if (targetConfig.isEnabled()) {
-                Output output = new Output(
-                        targetName,
-                        targetConfig.getTarget(),
-                        Integer.parseInt(targetConfig.getPort()),
-                        routingConfig.getConnectionTimeout(),
-                        routingConfig.getReadTimeout(),
-                        routingConfig.getWriteTimeout(),
-                        routingConfig.getReconnectInterval(),
-                        metricRegistry
-                );
-                this.outputMap.put(targetName, output);
+                String hostname = targetConfig.getTarget();
+                int port = Integer.parseInt(targetConfig.getPort());
+                int connectionTimeout = routingConfig.getConnectionTimeout();
+                // FIXME blocked ~ you ran it in the constructor not in the method
+                InetSocketAddress isa = new InetSocketAddress(hostname, port);
+                // TODO count connectLatency
+                LOGGER.debug("opening client to isa <[{}]>", isa);
+                CompletableFuture<RelpClient> futureClient = clientFactory
+                        .open(isa)
+                        .orTimeout(connectionTimeout, TimeUnit.MILLISECONDS);
+
+                futureClient.thenAccept(client -> {
+                    String offer = ("\nrelp_version=0\nrelp_software=cfe_35\ncommands=" + "syslog" + "\n");
```

Comment:
Do we want to have this as variable in code instead of defined field variable or something? Also why string concatenation as no variables are present